### PR TITLE
Rework wait-wake API for smart-contracts

### PR DESCRIPTION
**Pull request by shamilsan**
- Add `gr_wake` syscall.
- Move the message to the wait list after calling `gr_wait`.
- Move the message from the wait list to the main queue after calling `gr_wake`.

Done:
- [x] Fix node tests
- [x] Fix RPC test

@gear-tech/dev

### CODE REVIEW

File: core-backend/src/funcs.rs [hunk 1]
```diff
@@ -226,7 +226,18 @@ pub(crate) fn wait<E: Ext>(ext: LaterExt<E>) -> impl Fn() -> Result<(), &'static
     move || {
         let _ = ext.with(|ext: &mut E| ext.wait())?;
         // Intentionally return an error to break the execution
-        Err("wait")
+        Err("exit")
```
Comment (NikVolf) [inline core-backend/src/funcs.rs:229]: let's make `const EXIT_TRAP_STR: &'static str = "exit"` ?

**Comment by shamilsan** [comment]
Problem seems to be here:

https://github.com/gear-tech/gear/blob/55e3be3cc1c807e193fb7dc4a652f0957ea80271/gtest/src/runner.rs#L76

Will be fixed within #109.

File: core-backend/src/funcs.rs (hunk 1 above)
Comment (shamilsan) [reply core-backend/src/funcs.rs]: Done.

**Review by ukint-vs — APPROVED** [review]

File: core/src/storage.rs [hunk 2]
```diff
@@ -173,12 +173,12 @@ impl InMemoryWaitList {
 }
 
 impl WaitList for InMemoryWaitList {
-    fn insert(&mut self, waker_id: MessageId, message: Message) {
-        self.inner.insert(waker_id, message);
+    fn insert(&mut self, id: MessageId, message: Message) {
```
Comment (NikVolf) [inline core/src/storage.rs:176]: `WaitList` should be partitioned by `program_id`

Comment (shamilsan) [reply core/src/storage.rs:176]: Ok, I'll add an issue.

**Pull request merged**